### PR TITLE
Add a helper function to customize K8s addon yamls and use it to customize Calico addons on GKE

### LINKED ISSUE #65045: Fail to run Calico 3.0+ with existing addon yaml files under cluster/addons/calico-policy-controller/
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:

The K8s cluster with NETWORK_POLICY_PROVIDER=calico could not be brought up in GCE when using Calico 3.0+ release. It would complain a number of issues, e.g., missing CRDs, authentication issues, ...

**What you expected to happen**:

K8s cluster can be brought up with Calico 3.0+ release

**How to reproduce it (as minimally and precisely as possible)**:

- Change image versions in cluster/addons/calico-policy-controller/{calico-node-daemonset, typha-depolyment}.yaml to Calico 3.0+ release

- make quick-release

- set ENV vars

export KUBE_CLUSTER_INITIALIZATION_TIMEOUT=3600
export NUM_NODES=1
export KUBE_UP_AUTOMATIC_CLEANUP=true
export KUBE_GCE_ZONE=us-west1-a
export KUBE_GCE_ENABLE_IP_ALIASES=true
export NETWORK_POLICY_PROVIDER=calico

- run cluster/kube-up.sh to spin up a K8s cluster on GCE

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration: gce
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by jingax10**
**What this PR does / why we need it**:

Allow customizing Calico addon in GCP. With #65022, this allows us to do a couple of things:, e.g., run Calico 3.0+ on GCP, use a non-default MTU etc.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #65045, #65067

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -2436,6 +2453,9 @@ EOF
   if [[ "${NETWORK_POLICY_PROVIDER:-}" == "calico" ]]; then
     setup-addon-manifests "addons" "calico-policy-controller"
 
+    setup-addon-custom-yaml "addons" "calico-policy-controller" "calico-node-daemonset.yaml" "${CUSTOM_CALICO_NODE_DAEMONSET_YAML}"
```
Comment (thockin) [inline cluster/gce/gci/configure-helper.sh]: where is CUSTOM_CALICO_NODE_DAEMONSET_YAML defined?

Comment (jingax10) [reply cluster/gce/gci/configure-helper.sh]: Changed to refer it as ${CUSTOM_CALICO_NODE_DAEMONSET_YAML:-} to allow it is unbound if not specified.

File: cluster/gce/gci/configure-helper.sh [hunk 2]
```diff
@@ -2286,6 +2286,23 @@ EOF
...
+# $2: manifest source dir
+# $2: manifest file
+# $3: custom yaml
+function setup-addon-custom-yaml {
+  local -r manifest_path="/etc/kubernetes/$1/$2/$3"
+  local -r custom_yaml="$4"
+  if [ -n "${custom_yaml:-}" ]; then
+    # Replace with custom manifest.
+    cat > "${manifest_path}" <<EOF
+$(echo "${custom_yaml}")
```
Comment (thockin) [inline cluster/gce/gci/configure-helper.sh]: Why is this execing echo?  Why not simply reference the variable?

File: cluster/gce/gci/configure-helper.sh [hunk 3]
```diff
@@ -2286,6 +2286,23 @@ EOF
   fi
 }
 
+# A helper function to set up a custom yaml for a k8s addon.
```
Comment (thockin) [inline cluster/gce/gci/configure-helper.sh]: A couple real examples would help

**Review by thockin — COMMENTED** [review]
Not sure I get what this is doing - is it overwriting existing files?

File: cluster/gce/gci/configure-helper.sh (hunk 3 above)
Comment (jingax10) [reply cluster/gce/gci/configure-helper.sh]: It is invoked below in line 2456 & 2457.

File: cluster/gce/gci/configure-helper.sh (hunk 2 above)
Comment (jingax10) [reply cluster/gce/gci/configure-helper.sh]: It does not need echo cmd and verified it. I just copied line 2261 $(echo "$CUSTOM_KUBE_DNS_YAML"). I will send another PR to clean it up.

**Review by jingax10 — COMMENTED** [review]
Tim, thanks for the review! PTAL

**Comment by thockin** [comment]
@roberthbailey Are you more familiar with this code?  I mean, it looks fine, but I don't know the norms of this any more...

File: cluster/gce/gci/configure-helper.sh [hunk 4]
```diff
@@ -2286,6 +2286,23 @@ EOF
...
+#
+# $1: addon category under /etc/kubernetes
+# $2: manifest source dir
+# $2: manifest file
```
Comment (roberthbailey) [inline cluster/gce/gci/configure-helper.sh]: nit: this is $3 and the following line is $4

**Comment by roberthbailey** [comment]
This change looks ok to me.

/cc @johnsonj

File: cluster/gce/gci/configure-helper.sh (hunk 4 above)
Comment (jingax10) [reply cluster/gce/gci/configure-helper.sh]: Good catch! Fixed.

**Review by jingax10 — COMMENTED** [review]
Thanks @roberthbailey! PTAL

**Pull request merged**